### PR TITLE
Clause checks for combined constructs

### PR DESCRIPTION
**Pull request by DavidTruby**
This includes a refactor of the existing combined construct checks
that were present, as well as adding the remaining combined constructs
that had not been implemented yet.

### CODE REVIEW

File: test/semantics/omp-clause-validity01.f90 [hunk 1]
```diff
@@ -445,7 +445,6 @@
   enddo
   !$omp end taskloop simd
 
-  !ERROR: REDUCTION clause is not allowed on the TASKLOOP SIMD directive
   !$omp taskloop simd reduction(+:a)
```
Comment (ichoyjx) [inline test/semantics/omp-clause-validity01.f90]: No reduction clause can be specified. (4.5 Spec P92)

File: lib/semantics/check-omp-structure.cpp [hunk 2]
```diff
@@ -409,6 +556,15 @@ void OmpStructureChecker::Enter(const parser::OpenMPBlockConstruct &x) {
   case parser::OmpBlockDirective::Directive::Master:
     PushContext(beginDir.source, OmpDirective::MASTER);
     break;
+    // 2.11.5 target-parallel-clause -> target-clause |
+    //                                  parallel-clause
+  case parser::OmpBlockDirective::Directive::TargetParallel: {
+    PushContext(beginDir.source, OmpDirective::TARGET_PARALLEL);
+    SetContextAllowed(
+        (targetAllowedClauses | parallelAllowedClauses) - OmpClause::COPYIN);
```
Comment (ichoyjx) [inline lib/semantics/check-omp-structure.cpp]: Could you add a small test for `copyin` clause? Thx.

File: test/semantics/omp-clause-validity01.f90 (hunk 1 above)
Comment (DavidTruby) [reply test/semantics/omp-clause-validity01.f90]: Interesting. The reduction clause actually _can_ be specified in 5.0, but not in 4.5. Which is our "source of truth" here? I assume 4.5, in which case I'll add this limitation back in.

**Comment by DavidTruby** [comment]
@ichoyjx Sorry for letting this rot for so long! I've added some more tests including the one you requested, and I've added back the restriction on reduction clauses on taskloop simd (and the relevant test).

**Pull request closed without merging**